### PR TITLE
Add params.permit! which raises on unpermitted params

### PR DESCRIPTION
**Pull request by martinemde**
### Motivation / Background

`action_on_unpermitted_parameters` only allows global control of something that often needs to be controlled on a per-action basis.

### Detail

This Pull Request introduces `params.permit!` which raises if unpermitted parameters are present. It also accepts a block which yields the array of unpermitted keys instead of raising, allowing for custom handling or logging. 

### Additional information

This was the most obvious part of #51674 that could be merged. It lays the groundwork for everything else in that PR, making the remaining code much easier to understand and review.

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: actionpack/lib/action_controller/metal/strong_parameters.rb [hunk 1]
```diff
@@ -646,9 +656,18 @@ def require(key)
     #     params.permit(person: { '0': [:email], '1': [:phone]}).to_h
     #     # => {"person"=>{"0"=>{"email"=>"none@test.com"}, "1"=>{"phone"=>"555-6789"}}}
     def permit(*filters)
-      params = self.class.new
+      permit_filters(filters, &method(:unpermitted_parameters))
+    end
+
+    # Same as permit, but raises on unpermitted parameters by default.
+    # Ignores ActionController::Parameters.action_on_unpermitted_parameters
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/metal/strong_parameters.rb:663]:
```suggestion
```

File: actionpack/lib/action_controller/metal/strong_parameters.rb [hunk 2]
```diff
@@ -646,9 +656,18 @@ def require(key)
     #     params.permit(person: { '0': [:email], '1': [:phone]}).to_h
     #     # => {"person"=>{"0"=>{"email"=>"none@test.com"}, "1"=>{"phone"=>"555-6789"}}}
     def permit(*filters)
-      params = self.class.new
+      permit_filters(filters, &method(:unpermitted_parameters))
+    end
+
+    # Same as permit, but raises on unpermitted parameters by default.
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/metal/strong_parameters.rb:662]:
```suggestion
    # Same as `permit`, but raises on unpermitted parameters.
```

File: actionpack/lib/action_controller/metal/strong_parameters.rb [hunk 3]
```diff
@@ -646,9 +656,18 @@ def require(key)
...
+    # Ignores ActionController::Parameters.action_on_unpermitted_parameters
+    #
+    # Accepts a block which will yield any unpermitted keys instead of raising.
+    def permit!(*filters, &block)
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/metal/strong_parameters.rb:666]: We already have a `permit!` method.

File: actionpack/lib/action_controller/metal/strong_parameters.rb [hunk 4]
```diff
@@ -646,9 +656,18 @@ def require(key)
...
+      block ||= method(:raise_unpermitted_parameters)
+      permit_filters(filters, &block)
+    end
 
-      filters.flatten.each do |filter|
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/metal/strong_parameters.rb:651]: This code is wrong now

File: actionpack/lib/action_controller/metal/strong_parameters.rb (hunk 3 above)
Comment (martinemde) [reply actionpack/lib/action_controller/metal/strong_parameters.rb:666]: uh oh, haha. Of course. When we were talking about `expect!` in the other PR, I completely forgot about this. Hmmmm...

**Comment by martinemde** [comment]
Oops, I completely forgot about there being a permit! already. Maybe this is a non-starter or we go back to `permit_only`. Shoot. Sorry for the extra noise here. I thought this would simplify things but I think I've made it more complex. I can just close this for now and we'll come back to it.

**Pull request closed without merging**